binary_search_tree/tree.py

```diff
@@ -1,59 +1,163 @@
 class TreeNode:
-    def __init__(self, key, val = None):
-        if val == None:
+    def __init__(self, key, val=None):
+        if val is None:
             val = key
 
         self.key = key
         self.value = val
         self.left = None
         self.right = None
-        
+
+    def to_json(self):
+        return {
+            "key": self.key,
+            "value": self.value
+        }
 
 
 class Tree:
     def __init__(self):
         self.root = None
 
-    # Time Complexity: 
-    # Space Complexity: 
-    def add(self, key, value = None):
-        pass
+    # ----------- add ----------
+
+    # Time Complexity: O(log n) for balanced BST, O(n) for unbalanced
+    # Space Complexity: O(1)
+
+    def add_helper(self, current, key, value):
+        if current is None:
+            return TreeNode(key, value)
+        if current.key >= key:
+            current.left = self.add_helper(current.left, key, value)
+        elif current.key < key:
+            current.right = self.add_helper(current.right, key, value)
+        return current
+
+    def add(self, key, value=None):
+        if self.root is None:
+            self.root = TreeNode(key, value)
+        else:
+            self.add_helper(self.root, key, value)
+
+    # ----------- find ----------
+
+    # Time Complexity: O(log n) for balanced BST, O(n) for unbalanced
+    # Space Complexity: O(1)
+
+    def find_helper(self, current, key):
```

Comment:
👍 However due to the recursive stack, the space complexity is O(log n) for a balanced tree.